### ISSUE TITLE
Question: Best practice for PUID and PGID

### ISSUE THREAD

**Issue by smokemyshoes** [authelia/authelia#1925]
Hi

I'm running Authelia and mariadb containers in a basic setup so far.  I'm not using redit and LDAP yet but they are both on my roadmap. I'm looking to improve security and would like to know what is considered best practice (from an opsec perspective) in relation to the PUID and PGID environment variable options.

Is it advised to map the PUID:PGID inside these two containers to a user:group on the host and if so:

1. Should it be a dedicated user:group for authelia and mariadb?
2. Should both containers be mapped to the same host user:group?
3. What are the required privileges for the user(s) on the host? i.e. which group(s) do they need to be part off?

I have been looking around but haven't been able to find any recommendations so I hope you can help me.

**Comment by james-d-elliott** [comment]
I don't know if there is an arbitrary best practice but I will explain some of my personal rules and those that I implemented at work for container FS as well as PUID/PGID and the reasons behind it. 

It's a bit wordy but unless I convey why I do something I don't think there is any sense in someone else doing so. You have to evaluate my system and see which parts suit you or are desirable to you.

### The UID

So I have always taken the approach that the UID of containers and the underlying filesystems must fit one of two categories:

1. They have a high number that should never be assigned to a user. I usually do this option and all container files are numbers in the 7xxx range.
2. They are reserved as users so they cannot be reused by actual user accounts.

The reason for this is you never want a normal user accessing files they are not meant to access. I think this is especially important when it comes to secrets if they are mounted as files.

### The GID

The rules I use for GID are extremity similar to those above. The only alteration is obviously when containers are meant to have access to shared files. I prefer this to giving a o+rw or o+r chmod. I generally use the 8xxx range, but any system works as long as it's not going to be used by normal groups.

### The Container

This is probably the most difficult one to implement and probably the least important. However, container runtimes allow you to specify the UID and GID of the user running the entrypoint. 

To me this is desirable as this means the container never runs as root. The issue is ***many if not most*** containers don't really allow you to do this in a nice way (errors, or you have to override scripts, etc), as they do things only root can do by default without specific additional permissions containers usually don't have.

Authelia actually supports this. So you can run the container in an unprivileged manner and never allow it access to root, just by specifying the UID and GID in docker compose. For example to run Authelia as UID 7001 and GID 8005 your docker-compose would look like this:

```yaml
services:
  authelia:
    user: 7001:8005
```

The complication with implementing it this way is you manually have to configure the permissions of files. I'd recommend something like the following if you plan to implement something similar to this:

```bash
chown -r 7001:8005 ./data/authelia
find ./.data/authelia/ -type d -exec chmod 750 {} \;
find ./.data/authelia/ -type f -exec chmod 640 {} \;
```

**Comment by smokemyshoes** [comment]
@james-d-elliott Thanks for your thorough reply, and thanks for clarifying that `user: XXXX:YYYY` and `PUID:XXXX`+ `PGID:YYYY` actually does different stuff. I missed that part.

I assume that even if I go the `PUID:XXXX`+ `PGID:YYYY` route, I will have to `chown -r ....` the volumes if I create them in advance, correct?

**Comment by james-d-elliott** [comment]
> 
> 
> @james-d-elliott Thanks for your thorough reply, and thanks for clarifying that `user: XXXX:YYYY` and `PUID:XXXX`+ `PGID:YYYY` actually does different stuff. I missed that part.
> 
> I assume that even if I go the `PUID:XXXX`+ `PGID:YYYY` route, I will have to `chown -r ....` the volumes if I create them in advance, correct?

If you just use PUID and PGID it should chown them for you. Because of the entrypoint will hit these lines:

https://github.com/authelia/authelia/blob/master/entrypoint.sh#L8-L9
